src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,170 @@
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E element, Node<E> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+
+        Node(E element) {
+            this.item = element;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> newNode = new Node<>(null, value,null);
+            head = tail = newNode;
+        } else {
+            Node<T> newNode = new Node<>(value);
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            if (size == 0) {
+                head = tail = newNode;
+            } else {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            }
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> current = getNode(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = getNode(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNode(index);
+        if (nodeToRemove == head) {
+            head = head.next;
+            if (head != null) {
+                head.prev = null;
+            }
+        } else if (nodeToRemove == tail) {
+            tail = tail.prev;
+            if (tail != null) {
+                tail.next = null;
+            }
+        } else {
+            nodeToRemove.prev.next = nodeToRemove.next;
+            nodeToRemove.next.prev = nodeToRemove.prev;
+        }
+        size--;
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.item == null && object == null || (current.item != null
```

Comment:
Be careful with comparing objects for equality using '==' as it compares references, not values. Also, consider the case where 'null' can be passed as a method parameter.